airflow/providers/amazon/aws/operators/s3.py

```diff
@@ -503,23 +509,37 @@ def __init__(
         self.bucket = bucket
         self.keys = keys
         self.prefix = prefix
+        self.from_datetime = from_datetime
+        self.to_datetime = to_datetime
         self.aws_conn_id = aws_conn_id
         self.verify = verify
 
         self._keys: str | list[str] = ""
 
-        if not exactly_one(prefix is None, keys is None):
-            raise AirflowException("Either keys or prefix should be set.")
+        if not exactly_one(keys is None, all([prefix is None, from_datetime is None, to_datetime is None])):
+            raise AirflowException(
+                "Either keys or at least one of prefix, from_datetime, to_datetime should be set."
```

Comment:
That message does not match what the code is enforcing.  The code says either keys or all of the others, no? Isn't this saying either keys can be empty or all the rest are?  Maybe I'm just getting confused by the double negative here.

What is you change it to a positive check like:


```suggestion
        if not exactly_one(keys, at_least_one(prefix, from_datetime, to_datetime])):
```

Reply:
I was playing with that in ipython and looking at the unit tests, and apparently an empty list and an empty string are considered valid values?   That feels wrong to me.  If one must be provided, then surely providing an empty value shouldn't count?   That complicates things and I guess explains why this statement is convoluted.

My initial statement stands though, I'm pretty sure this is not testing what your error message says it is.

Reply:
I agree that it will be easier to understand without the negative sections.
do you suggest to add the exactlyone function to airflow helpers?
The statement checks if keys are given or at least one of the other params were given.
it could be only the keys or any combination that includes one of the other params

about empty list or empty string they are both valid options for keys, maybe someone has this operator in a dag that gets dynamically which keys to delete but for a specific run he didn't have any keys to delete so the value of keys will be the empty list and the action that will be performed is to not delete any key